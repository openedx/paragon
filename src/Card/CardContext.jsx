import React, { createContext } from 'react';
import PropTypes from 'prop-types';

const CardContext = createContext({});

function CardContextProvider({
  orientation,
  children,
  isLoading,
  variant,
}) {
  return (
    <CardContext.Provider value={{ orientation, isLoading, variant }}>
      {children}
    </CardContext.Provider>
  );
}

CardContextProvider.propTypes = {
  /** Specifies which orientation to use. */
  orientation: PropTypes.oneOf(['horizontal', 'vertical']),
  /** Specifies loading state. */
  isLoading: PropTypes.bool,
  /** Specifies content of the component. */
  children: PropTypes.node,
  /** Specifies `Card` style variant */
  variant: PropTypes.oneOf(['light', 'dark', 'muted']),
};

CardContextProvider.defaultProps = {
  orientation: 'vertical',
  isLoading: false,
  children: null,
  variant: 'light',
};

export { CardContextProvider };
export default CardContext;
