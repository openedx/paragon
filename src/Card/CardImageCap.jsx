import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Skeleton from 'react-loading-skeleton';
import CardContext from './CardContext';
import cardSrcFallbackImg from './fallback-default.png';

const SKELETON_HEIGHT_VALUE = 140;
const LOGO_SKELETON_HEIGHT_VALUE = 41;

const CardImageCap = React.forwardRef(({
  src,
  fallbackSrc,
  srcAlt,
  logoSrc,
  fallbackLogoSrc,
  logoAlt,
  skeletonHeight,
  skeletonWidth,
  logoSkeleton,
  logoSkeletonHeight,
  logoSkeletonWidth,
  className,
}, ref) => {
  const { orientation, isLoading } = useContext(CardContext);
  const [showImageCap, setShowImageCap] = useState(false);
  const [showLogoCap, setShowLogoCap] = useState(false);

  const wrapperClassName = `pgn__card-wrapper-image-cap ${orientation}`;

  if (isLoading) {
    return (
      <div
        className={classNames(wrapperClassName, className)}
        data-testid="image-loader-wrapper"
      >
        <Skeleton
          containerClassName="pgn__card-image-cap-loader"
          height={orientation === 'horizontal' ? '100%' : skeletonHeight}
          width={skeletonWidth}
        />
        {logoSkeleton && (
          <Skeleton
            containerClassName="pgn__card-logo-cap"
            height={logoSkeletonHeight}
            width={logoSkeletonWidth}
          />
        )}
      </div>
    );
  }

  const handleSrcFallback = (event, altSrc, imageKey) => {
    const { currentTarget } = event;

    if (!altSrc || currentTarget.src.endsWith(altSrc)) {
      if (imageKey === 'imageCap') {
        currentTarget.src = cardSrcFallbackImg;
      } else {
        setShowLogoCap(false);
      }

      return;
    }

    currentTarget.src = altSrc;
  };

  return (
    <div className={classNames(className, wrapperClassName)} ref={ref}>
      {!!src && (
        <img
          className={classNames('pgn__card-image-cap', { show: showImageCap })}
          src={src}
          onError={(event) => handleSrcFallback(event, fallbackSrc, 'imageCap')}
          onLoad={() => setShowImageCap(true)}
          alt={srcAlt}
        />
      )}
      {!!logoSrc && (
        <img
          className={classNames('pgn__card-logo-cap', { show: showLogoCap })}
          src={logoSrc}
          onError={(event) => handleSrcFallback(event, fallbackLogoSrc, 'logoCap')}
          onLoad={() => setShowLogoCap(true)}
          alt={logoAlt}
        />
      )}
    </div>
  );
});

CardImageCap.propTypes = {
  /** Specifies class name to append to the base element. */
  className: PropTypes.string,
  /** Specifies image src. */
  src: PropTypes.string,
  /** Specifies fallback image src. */
  fallbackSrc: PropTypes.string,
  /** Specifies image alt text. */
  srcAlt: PropTypes.string,
  /** Specifies logo src to put on top of the image. */
  logoSrc: PropTypes.string,
  /** Specifies fallback image logo src. */
  fallbackLogoSrc: PropTypes.string,
  /** Specifies logo image alt text. */
  logoAlt: PropTypes.string,
  /** Specifies height of Image skeleton in loading state. */
  skeletonHeight: PropTypes.number,
  /** Specifies width of Image skeleton in loading state. */
  skeletonWidth: PropTypes.number,
  /** Specifies whether the cap should be displayed during loading. */
  logoSkeleton: PropTypes.bool,
  /** Specifies height of Logo skeleton in loading state. */
  logoSkeletonHeight: PropTypes.number,
  /** Specifies width of Logo skeleton in loading state. */
  logoSkeletonWidth: PropTypes.number,
};

CardImageCap.defaultProps = {
  src: undefined,
  fallbackSrc: cardSrcFallbackImg,
  logoSrc: undefined,
  fallbackLogoSrc: undefined,
  className: undefined,
  srcAlt: undefined,
  logoAlt: undefined,
  skeletonHeight: SKELETON_HEIGHT_VALUE,
  logoSkeleton: false,
  logoSkeletonHeight: LOGO_SKELETON_HEIGHT_VALUE,
  skeletonWidth: undefined,
  logoSkeletonWidth: undefined,
};

export default CardImageCap;
