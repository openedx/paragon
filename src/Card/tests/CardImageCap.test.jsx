import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import CardImageCap from '../CardImageCap';
import CardContext from '../CardContext';

// eslint-disable-next-line react/prop-types
function CardImageCapWrapper({ orientation = 'vertical', isLoading, ...props }) {
  return <CardContext.Provider value={{ orientation, isLoading }}><CardImageCap {...props} /></CardContext.Provider>;
}

describe('<CardImageCap />', () => {
  it('renders with scr prop and srcAlt', () => {
    const tree = renderer.create((
      <CardImageCapWrapper src="http://fake.image" srcAlt="Alt text" />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with src, logoSrc and logoAlt props', () => {
    const tree = renderer.create((
      <CardImageCapWrapper src="http://fake.image" logoSrc="http://fake.image" logoAlt="Logo alt" />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with scr prop in horizontal orientation', () => {
    const tree = renderer.create((
      <CardImageCapWrapper orientation="horizontal" src="http://fake.image" />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with src and logoSrc prop in horizontal orientation', () => {
    const tree = renderer.create((
      <CardImageCapWrapper orientation="horizontal" src="http://fake.image" logoSrc="http://fake.image" />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with correct horizontal class', async () => {
    const { container } = render(<CardImageCapWrapper orientation="horizontal" src="http://fake.image" />);

    expect(container.firstChild).toHaveClass('pgn__card-wrapper-image-cap horizontal');
  });

  it('renders with correct alt texts', async () => {
    render(
      <CardImageCapWrapper
        src="http://src.image/"
        logoSrc="http://logo.image/"
        srcAlt="Src alt text"
        logoAlt="Logo alt text"
      />,
    );

    expect(screen.getByAltText('Src alt text').src).toEqual('http://src.image/');
    expect(screen.getByAltText('Logo alt text').src).toEqual('http://logo.image/');
  });

  it('render with loading state', () => {
    render(
      <CardImageCapWrapper
        src="http://fake.image"
        logoSrc="http://fake.image"
        srcAlt="Src alt text"
        logoAlt="Logo alt text"
        isLoading
        logoSkeleton
      />,
    );

    expect(screen.queryByAltText('Src alt text')).toBeNull();
    expect(screen.queryByAltText('Logo alt text')).toBeNull();

    expect(screen.getByTestId('image-loader-wrapper')).toBeInTheDocument();
    expect(screen.getByTestId('image-loader-wrapper').firstChild).toHaveClass('pgn__card-image-cap-loader');
    expect(screen.getByTestId('image-loader-wrapper').lastChild).toHaveClass('pgn__card-logo-cap');
  });

  it('replaces image with fallback one in case of error', () => {
    render(
      <CardImageCapWrapper
        src="http://src.image/"
        srcAlt="Src alt text"
        fallbackSrc="http://src.image.fallback/"
        logoSrc="http://logo.image/"
        fallbackLogoSrc="http://logo.image.fallback/"
        logoAlt="Logo alt text"
      />,
    );
    const srcImg = screen.getByAltText('Src alt text');
    expect(srcImg.src).toEqual('http://src.image/');
    fireEvent.load(srcImg);
    fireEvent.error(srcImg);
    expect(srcImg.src).toEqual('http://src.image.fallback/');

    const logoImg = screen.getByAltText('Logo alt text');
    expect(logoImg.src).toEqual('http://logo.image/');
    fireEvent.load(srcImg);
    fireEvent.error(logoImg);
    expect(logoImg.src).toEqual('http://logo.image.fallback/');
  });

  it('hiding component if it isn`t fallbackLogoSrc and logoSrc don`t work', () => {
    render(<CardImageCapWrapper logoSrc="fakeURL" logoAlt="Logo alt text" />);

    const logoImg = screen.getByAltText('Logo alt text');
    fireEvent.load(logoImg);
    expect(logoImg.className).toEqual('pgn__card-logo-cap show');
    fireEvent.error(logoImg);
    expect(logoImg.className).toEqual('pgn__card-logo-cap');
  });

  it('hiding component if it isn`t fallbackSrc and src don`t work', () => {
    render(<CardImageCapWrapper src="fakeURL" fallbackSrc="fakeURL" srcAlt="Src alt text" />);

    const srcImg = screen.getByAltText('Src alt text');
    fireEvent.load(srcImg);
    fireEvent.error(srcImg);
    // test-file-stub is what our fileMock.js returns for all images
    expect(srcImg.src.endsWith('test-file-stub')).toEqual(true);
  });
});
