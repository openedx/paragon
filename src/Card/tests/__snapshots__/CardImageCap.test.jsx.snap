// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardImageCap /> renders with scr prop and srcAlt 1`] = `
<div
  className="pgn__card-wrapper-image-cap vertical"
>
  <img
    alt="Alt text"
    className="pgn__card-image-cap"
    onError={[Function]}
    onLoad={[Function]}
    src="http://fake.image"
  />
</div>
`;

exports[`<CardImageCap /> renders with scr prop in horizontal orientation 1`] = `
<div
  className="pgn__card-wrapper-image-cap horizontal"
>
  <img
    className="pgn__card-image-cap"
    onError={[Function]}
    onLoad={[Function]}
    src="http://fake.image"
  />
</div>
`;

exports[`<CardImageCap /> renders with src and logoSrc prop in horizontal orientation 1`] = `
<div
  className="pgn__card-wrapper-image-cap horizontal"
>
  <img
    className="pgn__card-image-cap"
    onError={[Function]}
    onLoad={[Function]}
    src="http://fake.image"
  />
  <img
    className="pgn__card-logo-cap"
    onError={[Function]}
    onLoad={[Function]}
    src="http://fake.image"
  />
</div>
`;

exports[`<CardImageCap /> renders with src, logoSrc and logoAlt props 1`] = `
<div
  className="pgn__card-wrapper-image-cap vertical"
>
  <img
    className="pgn__card-image-cap"
    onError={[Function]}
    onLoad={[Function]}
    src="http://fake.image"
  />
  <img
    alt="Logo alt"
    className="pgn__card-logo-cap"
    onError={[Function]}
    onLoad={[Function]}
    src="http://fake.image"
  />
</div>
`;
